cluster-autoscaler/expander/factory/expander_factory.go

```diff
@@ -31,30 +31,41 @@ import (
 	kube_client "k8s.io/client-go/kubernetes"
 )
 
-// ExpanderStrategyFromString creates an expander.Strategy according to its name
-func ExpanderStrategyFromString(expanderFlag string, cloudProvider cloudprovider.CloudProvider,
+// ExpanderStrategyFromStrings creates an expander.Strategy according to the names of the expanders passed in
+func ExpanderStrategyFromStrings(expanderFlags []string, cloudProvider cloudprovider.CloudProvider,
 	autoscalingKubeClients *context.AutoscalingKubeClients, kubeClient kube_client.Interface,
 	configNamespace string) (expander.Strategy, errors.AutoscalerError) {
-	switch expanderFlag {
-	case expander.RandomExpanderName:
-		return random.NewStrategy(), nil
-	case expander.MostPodsExpanderName:
-		return mostpods.NewStrategy(), nil
-	case expander.LeastWasteExpanderName:
-		return waste.NewStrategy(), nil
-	case expander.PriceBasedExpanderName:
-		if _, err := cloudProvider.Pricing(); err != nil {
-			return nil, err
+	var filters []expander.Filter
+	seenExpanders := map[string]struct{}{}
+	for _, expanderFlag := range expanderFlags {
+		if _, ok := seenExpanders[expanderFlag]; ok {
+			return nil, errors.NewAutoscalerError(errors.InternalError, "Expander %s was specified multiple times, each expander must not be specified more than once", expanderFlag)
+		}
+		seenExpanders[expanderFlag] = struct{}{}
+
+		switch expanderFlag {
+		case expander.RandomExpanderName:
+			filters = append(filters, random.NewFilter())
```

Comment:
Shouldn't it be an error to specify an expander that's guaranteed to provide a single option with a fallback? E.g. `random,least-waste` is not really a reasonable config.

Reply:
True that the configuration is not a reasonable one, but I felt that catching this case wasn't worth the effort as it would require disambiguating between expanders that

1. Always return one result and
2. Those that can return more than one.

My thinking is that since it won't create problems then we can keep the simplicity of allowing this. Happy to be overruled on this though.

Reply:
Here's what it would look like: https://github.com/ryanmcnamara/autoscaler/compare/rm/expander-chain...ryanmcnamara:rm/expander-chain-demo?expand=1

Reply:
Yup, something like this. I'm not insisting on this one, but since you're generating an error for one class of invalid configs (duplicated entries), I thought it may make sense to go a step further and disallow bad ordering as well.

Reply:
Ok addressed in this pr now!

Reply:
Neat, thanks! Maybe worth adding a comment about this behavior? May be surprising for anyone adding a new expander in the future. Please also squash commits before merging.